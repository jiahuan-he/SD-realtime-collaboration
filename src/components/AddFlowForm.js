import React from 'react';

const button = {
    backgroundColor: "white",
    border: "1px solid",
    color: "black",
    padding: "5px 12px",
    cursor: "pointer",
}

const inputInvalid = {
    border: "2px solid #ed6663"
}

export default class AddFlowForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            flowID:"",
            from: "",
            to: "",
        };
    }

    handleChangeName = (event) => {
        const id = event.target.value.trim()
        this.setState({ flowID: id});
    }

    handleChangeFrom = (event) => {
        this.setState({ from: event.target.value});
    }

    handleChangeTo = (event) => {
        this.setState({ to: event.target.value});
    }

    isValidFlowID = () => {
        if(!this.state.flowID) return false
        return this.props.flows.filter((flow => flow.id === this.state.flowID)).length === 0
        && this.props.stocks.filter((stock => stock.id === this.state.stockName)).length === 0
    }

    isValidFrom = () => {
        if(this.state.from){
            if(this.props.stocks.length === 0 || this.props.stocks.filter((stock => stock.id === this.state.from)).length === 0) return false
        }
        if(this.state.from && this.state.to && this.state.from === this.state.to) return false
        return true 
    }

    isValidTo = () => {
        if(this.state.to){
            if(this.props.stocks.length === 0 || this.props.stocks.filter((stock => stock.id === this.state.to)).length === 0) return false
        }
        if(this.state.from && this.state.to && this.state.from === this.state.to) return false
        return true 
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <label>
                    Flow Name
                <input 
                    style={this.isValidFlowID()?null:inputInvalid} 
                    type="text" 
                    value={this.state.flowID} 
                    onChange={
                        (event) => {
                            this.handleChangeName(event)
                        }
                    } 
                />
                </label>
                <label>
                    From
                <input 
                    style={this.isValidFrom()?null:inputInvalid} 
                    type="text" 
                    value={this.state.from} onChange={this.handleChangeFrom} />
                </label>
                <label>
                    To
                <input 
                    style={this.isValidTo()?null:inputInvalid} 
                    type="text" 
                    value={this.state.to} onChange={this.handleChangeTo}/>
                </label>

                <input type="button" value={"Add Flow"} style={button}
                    onClick={() => {
                        if(!this.isValidFlowID()
                        || !this.isValidFrom()
                        || !this.isValidTo()
                        ) return
                        if(this.state.from) this.props.addDependenciesToStockOrFlow(this.state.flowID, this.state.from)
                        if(this.state.to) this.props.addDependenciesToStockOrFlow(this.state.flowID, this.state.to)
                        this.props.addFlow(this.state.flowID, this.state.from, this.state.to) 
                        this.setState({
                            flowID:"",
                            from: "",
                            to: "",})
                    }}
                />
            </form>
        )
    }
}