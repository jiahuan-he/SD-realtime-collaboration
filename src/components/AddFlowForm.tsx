import React, {FormEvent, CSSProperties } from 'react';
import { I_Stock, I_Flow, I_Parameter } from '../model';

interface State{
    flowID: string,
    from: string,
    to:string,
}

interface Props{
    stocks: Array<I_Stock>,
    flows: Array<I_Flow>,
    parameters: Array<I_Parameter>,
    addDependenciesToStockOrFlow: (from:string, to:string)=>void,
    addFlow: (id:string, from:string, to:string)=>void,
    addCloud: (id:string, from:string, to:string)=>void,
    button: CSSProperties,
    inputInvalid: CSSProperties,
    highlightStock: (id:string) => void,
}

export default class AddFlowForm extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);
        this.state = {
            flowID:"",
            from: "",
            to: "",
        };
    }

    handleChangeName = (event: FormEvent<HTMLInputElement>) => {
        const id = event.currentTarget.value.trim()
        this.setState({ flowID: id});
    }

    handleChangeFrom = (event: FormEvent<HTMLInputElement>) => {
        this.setState({ from: event.currentTarget.value});
    }

    handleChangeTo = (event: FormEvent<HTMLInputElement>) => {
        this.setState({ to: event.currentTarget.value});
    }

    isValidFlowID = () => {
        if(!this.state.flowID) return false
        return this.props.flows.filter((flow => flow.id === this.state.flowID)).length === 0
        && this.props.stocks.filter((stock => stock.id === this.state.flowID)).length === 0
        && this.props.parameters.filter((parameter => parameter.name === this.state.flowID)).length === 0
    }

    isValidFrom = () => {
        if(this.state.from){
            if(this.props.stocks.length === 0 || this.props.stocks.filter((stock => stock.id === this.state.from)).length === 0) return false
        }
        if(this.state.from && this.state.to && this.state.from === this.state.to) return false
        return true 
    }

    isValidTo = () => {
        if(this.state.to){
            if(this.props.stocks.length === 0 || this.props.stocks.filter((stock => stock.id === this.state.to)).length === 0) return false
        }
        if(this.state.from && this.state.to && this.state.from === this.state.to) return false
        return true 
    }

    render() {
        return (
            <form>
                <label>
                    Flow Name
                <input 
                    style={this.isValidFlowID()?undefined:this.props.inputInvalid} 
                    type="text" 
                    value={this.state.flowID} 
                    onChange={
                        (event) => {
                            this.handleChangeName(event)
                        }
                    } 
                />
                </label>
                <label>
                    From
                <input 
                    style={this.isValidFrom()?undefined:this.props.inputInvalid} 
                    type="text" 
                    value={this.state.from} onChange={this.handleChangeFrom} />
                </label>
                <label>
                    To
                <input 
                    style={this.isValidTo()?undefined:this.props.inputInvalid} 
                    type="text" 
                    value={this.state.to} onChange={this.handleChangeTo}/>
                </label>

                <input type="button" value={"Add Flow"} style={this.props.button}
                    onClick={() => {
                        if(!this.isValidFlowID()
                        || !this.isValidFrom()
                        || !this.isValidTo()
                        ) return
                        if(this.state.from) this.props.addDependenciesToStockOrFlow(this.state.flowID, this.state.from)
                        if(this.state.to) this.props.addDependenciesToStockOrFlow(this.state.flowID, this.state.to)
                        this.props.addFlow(this.state.flowID, this.state.from, this.state.to) 
                        this.props.addCloud(this.state.flowID, this.state.from, this.state.to) 
                        this.setState({
                            flowID:"",
                            from: "",
                            to: "",})
                    }}
                />
            </form>
        )
    }
}