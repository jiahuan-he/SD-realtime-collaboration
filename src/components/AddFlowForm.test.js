
import React from 'react';
import { shallow } from 'enzyme';
import AddFlowForm from './AddFlowForm';
describe('AddFlowForm', () => {
    it('should render correctly in "debug" mode', () => {       
        const form = shallow(<AddFlowForm
            debug
        ></AddFlowForm>);
        expect(form).toMatchSnapshot();
    });
});