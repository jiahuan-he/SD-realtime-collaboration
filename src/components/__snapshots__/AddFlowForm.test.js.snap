// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFlowForm should render correctly in "debug" mode 1`] = `
<form>
  <label>
    Flow Name
    <input
      onChange={[Function]}
      style={
        Object {
          "border": "2px solid #ed6663",
        }
      }
      type="text"
      value=""
    />
  </label>
  <label>
    From
    <input
      onChange={[Function]}
      style={null}
      type="text"
      value=""
    />
  </label>
  <label>
    To
    <input
      onChange={[Function]}
      style={null}
      type="text"
      value=""
    />
  </label>
  <input
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "border": "1px solid",
        "color": "black",
        "cursor": "pointer",
        "padding": "5px 12px",
      }
    }
    type="button"
    value="Add Flow"
  />
</form>
`;

exports[`AddFlowForm should render correctly in "debug" mode 2`] = `
<form>
  <label>
    Flow Name
    <input
      onChange={[Function]}
      style={
        Object {
          "border": "2px solid #ed6663",
        }
      }
      type="text"
      value=""
    />
  </label>
  <label>
    From
    <input
      onChange={[Function]}
      style={null}
      type="text"
      value=""
    />
  </label>
  <label>
    To
    <input
      onChange={[Function]}
      style={null}
      type="text"
      value=""
    />
  </label>
  <input
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "border": "1px solid",
        "color": "black",
        "cursor": "pointer",
        "padding": "5px 12px",
      }
    }
    type="button"
    value="Add Flow"
  />
</form>
`;
